### DISCUSSION TITLE
Type after transformation

### DISCUSSION THREAD

**Discussion by perevezencev** [react-hook-form/react-hook-form#12324] [Q&A]
Hello everyone! I have a code like this:

```typescript
const scheme = z.object({
  date: z
    .string()
    .transform((value) => value)
    .pipe(z.coerce.date()),
})

const useForm = <TFieldValues extends FieldValues>(options: {
    scheme: z.Schema<TFieldValues>
}) => {
    const { scheme } = options
    
    return useRHForm<TFieldValues>({
        resolver: zodResolver(scheme),
    })
}

useForm({
    scheme, // Type 'string' is not assignable to type 'Date'.
})
```

Why is this happening? 
Should the type of `date` be `z.ZodDate`, not `z.ZodPipeline<z.ZodEffects<z.ZodString, string, string>, z.ZodDate>`?

[Playground][1]

  [1]: https://www.typescriptlang.org/play/?#code/JYWwDg9gTgLgBAbzgVwM4FMBi0RwIaooYBKAEtlCADRwwCeY6cmw6ANgCYBqeby6hAL5wAZlAi4ARFHR4AxjAC0ACwgQA1opE5JAKFCRYiOAC8IHYgIhsAbuihxhYiXEkABVRu2UA9DNTWdlCoPmYcegbQ8ABU+IQmouJSYXq66AAehvByEAB2qPAceDBMALymunBVcAB0BVDAuQDmABQAlJXVNTBQePneIC0tNrz8bXClAHxwI3zoHdW1YMCMLSY1OfZy6DVFJe0dujn58KhyyuggZaY1EABGAFboCi0InXvoVLqChxlZcMcCkQsDgJnAADwAFRY7G4owEcAyJVyHEIMM4PDmqEmLQgYBgwDyqAAXIhOlUzhcrqT1gBlc6XPBQ9FwrGTb7jKZkxaA+BISmXJjCcp4glE8lwCUyGDIKC5YFkCggZmsDHw7GvCWLfyBew08yWAK2ewtAVXNpfRY-b66XRoEGUTWLM2fOA+HxwSEMJgAcnqjSaPrgwEIuQg8AIqGATVyeDubCYMAgtG9cB9ABFiugfTUOUA

**Comment by BrendanC23** [comment]
See [this section of the Zod docs](https://zod.dev/?id=zodtype-with-zodeffects). You can pass a `TTransformedValues` generic to `useForm` which fixes the issue. See [this TypeScript Playground](https://www.typescriptlang.org/play/?ssl=25&ssc=2&pln=17&pc=1#code/JYWwDg9gTgLgBAbzgVwM4FMBi0RwIaooYBKAEtlCADRwwCeY6cmw6ANgCYBqeby6hAL5wAZlAi4ARFHR4AxjAC0ACwgQA1opE5JAKFCRYiOAC8IHYgIhsAbuihxhYiXEkABVRu2UA9DNTWdlCoPmYcegbQ8ABU+IQmouJSYXq66AAehvByEAB2qPAceDBMALymunBVcAB0BVDAuQDmABQAlJXVNTBQePneIC0tNrz8bXClAHxwI3zoHdW1YMCMLSY1OfZy6DVFJe0dujn58KhyyuggZaY1EABGAFboCi0InXvoVLqCh-SMcI0wMh4OV1oDgQAeP7oCAiOBnC5XSYAbl00LgEGBQJBN0xMGxUIYMLhCMu6BRujSmSicGOBSIWBwEzgEIAKix2NxRgI4BkSrkOIQOZweHNCOVhVyxTRWQBhPIldI4vp0GWs3r9HDoKX8Qh89ACoWsEXcwgAHxQAvQIka2uZyCtNty2smLQgYBgwDyqAAXIhOlVSVc-esAMrnS54NmS0W6mjrABa5lZRIAIta1RrUANtbGBJNvuMpv7FnT4Egg0xhOV3Z7vQG4A2ZDBkFBcgyyBQQNHjTqBDL5blFTBM31s1q+6hXW9Fot-IF7CHzJYArZ7C1K20vosft9KWhGZRXg3KzQfD44Cn-gByeqNJrXgGEXIQeAEVDAJq5PB3NhMGAQLQRJwNeqbFOg141IWQA).

```
export const useForm = <TFieldValues extends FieldValues = FieldValues, TContext = any, TTransformedValues extends FieldValues | undefined = undefined>(options: {
    scheme: z.Schema<TFieldValues, z.ZodTypeDef, TTransformedValues>
}) => {
    const { scheme } = options
    
    return useRHForm<TFieldValues, TContext, TTransformedValues>({
        resolver: zodResolver(scheme),
    })
}
```
